addOns/exim/src/main/java/org/zaproxy/addon/exim/har/HarImporter.java

```diff
@@ -19,59 +19,156 @@
+    private static final List<String> ACCEPTED_VERSIONS =
+            List.of(HttpHeader.HTTP11, HttpHeader.HTTP2, HttpHeader.HTTP10, HttpHeader.HTTP09);
+
+    protected static final String STATS_HAR_FILE_ERROR = "import.har.file.errors";
+
+    private static ExtensionHistory extHistory;
+
     private ProgressPaneListener progressListener;
     private boolean success;
-    private static ExtensionHistory extHistory;
 
     public HarImporter(File file) {
         this(file, null);
     }
 
     public HarImporter(File file, ProgressPaneListener listener) {
         this.progressListener = listener;
-        importHarFile(file);
+        HarLog log = null;
+        try {
+            log = new HarReader().readFromFile(file).getLog();
+            importHarLog(log);
+        } catch (HarReaderException e) {
+            LOGGER.warn(
+                    "Failed to read HAR file: {} \n {}", file.getAbsolutePath(), e.getMessage());
+            Stats.incCounter(ExtensionExim.STATS_PREFIX + STATS_HAR_FILE_ERROR);
+            success = false;
+            completed();
+            return;
+        }
         completed();
     }
 
-    static List<HttpMessage> getHttpMessages(HarLog log) throws HttpMalformedHeaderException {
+    public HarImporter(HarLog harLog, ProgressPaneListener listener) {
+        this.progressListener = listener;
+        importHarLog(harLog);
+        completed();
+    }
+
+    private void importHarLog(HarLog log) {
+        processMessages(log);
+        Stats.incCounter(ExtensionExim.STATS_PREFIX + STATS_HAR_FILE);
+        success = true;
+    }
+
+    private void processMessages(HarLog log) {
+        List<HttpMessage> messages = null;
+        int before = log.getEntries().size();
+        preProcessHarLog(log);
+        int adjustment = before - log.getEntries().size();
+        try {
+            messages = HarImporter.getHttpMessages(log);
+        } catch (HttpMalformedHeaderException e) {
+            LOGGER.warn("Failed to process HAR entries. {}", e.getMessage());
+            LOGGER.debug(e, e);
+            Stats.incCounter(ExtensionExim.STATS_PREFIX + STATS_HAR_FILE_ERROR);
+            completed();
+            return;
+        }
+        int count = adjustment;
+        for (HttpMessage msg : messages) {
+            if (msg == null) {
+                continue;
+            }
+            persistMessage(msg);
+            updateProgress(++count, msg.getRequestHeader().getURI().toString());
+        }
+    }
+
+    private HarLog preProcessHarLog(HarLog log) {
+        List<HarEntry> entries =
+                log.getEntries().stream()
+                        .filter(this::entryHasUsableHttpVersion)
+                        .filter(this::entryIsNotLocalPrivate)
+                        .collect(Collectors.toList());
+        for (HarEntry entry : entries) {
+            entry.getResponse().getHeaders().forEach(this::preProcessHarHeaders);
+        }
```

Comment:
I couldn't figure out a way to do the for part while processing the stream 😔

Reply:
Make `preProcessHarHeaders` return the entry itself and call it in a `map`.